tensorflow/lite/micro/kernels/ceva/quantize.cc

```diff
@@ -0,0 +1,221 @@
+  TF_LITE_ENSURE(context, output != nullptr);
+
+  // TODO(b/128934713): Add support for fixed-point per-channel quantization.
+  // Currently this only support affine per-layer quantization.
+  TF_LITE_ENSURE_EQ(context, output->quantization.type,
+                    kTfLiteAffineQuantization);
+  const auto* affine_quantization =
+      reinterpret_cast<TfLiteAffineQuantization*>(output->quantization.params);
+  TF_LITE_ENSURE(context, affine_quantization);
+  TF_LITE_ENSURE(context, affine_quantization->scale);
+  TF_LITE_ENSURE(context, affine_quantization->scale->size == 1);
+
+  TF_LITE_ENSURE(context, input->type == kTfLiteFloat32 ||
+                              input->type == kTfLiteInt16 ||
+                              input->type == kTfLiteInt8);
+  TF_LITE_ENSURE(context, output->type == kTfLiteUInt8 ||
+                              output->type == kTfLiteInt8 ||
+                              output->type == kTfLiteInt16 ||
+                              output->type == kTfLiteInt32);
+
+  if ((input->type == kTfLiteInt16 && output->type == kTfLiteInt8) ||
+      (input->type == kTfLiteInt8 && output->type == kTfLiteInt8) ||
+      (input->type == kTfLiteInt8 && output->type == kTfLiteInt32) ||
+      (input->type == kTfLiteInt16 && output->type == kTfLiteInt16) ||
+      (input->type == kTfLiteInt16 && output->type == kTfLiteInt32)) {
+    double effective_scale = static_cast<double>(input->params.scale) /
+                             static_cast<double>(output->params.scale);
+
+    QuantizeMultiplier(effective_scale, &data->requantize_output_multiplier,
+                       &data->requantize_output_shift);
+  }
+
+  data->quantization_params.zero_point = output->params.zero_point;
+  data->quantization_params.scale = static_cast<double>(output->params.scale);
+
+  data->input_zero_point = input->params.zero_point;
+  return kTfLiteOk;
+}
+
+TfLiteStatus EvalCEVA(TfLiteContext* context, TfLiteNode* node) {
+  TFLITE_DCHECK(node->user_data != nullptr);
+
+  auto* data = static_cast<OpDataQuantizeReference*>(node->user_data);
+
+  const TfLiteEvalTensor* input = tflite::micro::GetEvalInput(context, node, 0);
+  TfLiteEvalTensor* output = tflite::micro::GetEvalOutput(context, node, 0);
+
+  if (input->type == kTfLiteFloat32) {
+    switch (output->type) {
+      case kTfLiteInt8: {
+        const float* input_data = tflite::micro::GetTensorData<float>(input);
+        int8_t* output_data = tflite::micro::GetTensorData<int8_t>(output);
+        const int flat_size =
+            MatchingFlatSize(tflite::micro::GetTensorShape(input),
+                             tflite::micro::GetTensorShape(output));
+
+#ifdef MCPS_MEASUREMENT
+        MCPS_START_ONE;
+#endif
+        CEVA_TFLM_AffineQuantize_Int8(input_data, output_data, flat_size,
+                                      data->quantization_params.scale,
+                                      data->quantization_params.zero_point);
```

Comment:
since this is the only case where you have an optimized implementation, it might be lower maintenance for you to have something like:

```cc
if (input->type == Float32 && output->type == INt8) {
// call optimized code
} else {
  return EvalQuantizeReference(...);
}
```

thoughts?

Reply:
Sounds reasonable, I'll make some changes later today.

Reply:
Done.